### PR TITLE
feat(tui): @file mention picker with frecency ranking (closes #570)

### LINKED ISSUE #570: feat(tui): file selection mentions with frecency ranking
### Reason

When composing prompts in the jcode TUI, there is no way to quickly reference files by name. Users have to manually type full file paths, which is error-prone and slow, especially in large codebases. This feature adds an autocomplete-style file mention system (similar to `@file` mentions in editors and chat apps) so users can quickly insert file references into their prompts.

### What was done

Implemented file selection mentions in the TUI composer with the following commits:

1. **`feat(tui): add file selection mentions`** — Core feature: typing `@` in the composer triggers a file mention popup. Files are indexed from the working directory (respecting `.gitignore` and common ignore patterns). New modules:
   - `file_mention.rs` (~1900 lines) — file index, fuzzy matching, and mention popup logic
   - `char_bag.rs` — compact character-bag fuzzy matching helper
   - `state_ui_input_helpers.rs` — input state helpers for mention navigation

2. **`fix(tui): preserve chat selection across composer drags`** — Ensures the chat view selection is not lost when dragging/resizing the composer, which previously caused mention state corruption.

3. **`fix(tui): harden file mention index and path matching`** — Hardened the file index against race conditions (files changing during indexing) and improved path matching accuracy (case-insensitive, partial path segments).

4. **`feat(tui): rank file mentions by frecency`** — Added frecency (frequency + recency) ranking so recently and frequently opened files appear first in the mention popup. Introduced a `FrecencyEntry` struct to track access patterns.

### Files changed (15 files, +3808 / -986 lines)

Key new files:
- `crates/jcode-tui/src/tui/app/file_mention.rs` — main mention logic
- `crates/jcode-tui/src/tui/app/char_bag.rs` — fuzzy matching
- `crates/jcode-tui/src/tui/app/state_ui_input_helpers.rs` — UI helpers

Modified files:
- `crates/jcode-tui/src/tui/app/input.rs` — `@` trigger handling
- `crates/jcode-tui/src/tui/ui_input.rs` — popup rendering and key handling
- `crates/jcode-tui/src/tui/app/tui_lifecycle.rs`, `tui_state.rs`, `mod.rs` — wiring

### Branch

The full implementation is available on my fork:

🔗 **https://github.com/alecuba16/jcode/tree/select_files**

### PR DESCRIPTION
**Pull request by alecuba16**
Closes #570.

## Problem

When composing prompts in the TUI there is no way to quickly reference files by name. Users must type full paths by hand, which is slow and error-prone in large codebases. #570 tracks this; maintainer guidance there suggested starting with project-local tracked files and a deterministic recency/frequency score, which is what this PR implements.

## What it does

Typing `@` in the composer opens a file-completion popover ranked by frecency (frequency + exponential recency decay). Accepting a completion replaces the `@query` with the path and records an inline file chip; backspace on a chip deletes the whole path token instead of one character.

- **Indexing**: `git ls-files --cached --others --exclude-standard` as the primary source, with a `walkdir` fallback honoring `.gitignore` for non-git directories. Rebuilt in the background with adaptive TTL (30s base, 4x for large workspaces) and a `max_files` safety cap. A `notify`-based watcher marks the index dirty on filesystem changes.
- **Matching**: case-insensitive substring matching behind a compact CharBag pre-filter, then a regex pass for path-segment matching. Gitignored directories are scanned lazily via `fs::read_dir` only when a query actually targets them.
- **Popover UX**: results grouped into Recent / Files sections. The selection never rests on a section header (the highlight and the Enter-accept path always resolve to the same real row), and the building-index hint row is a non-acceptable sentinel so Enter can never insert it as a chip.
- **Config**: tunable via `[file_mention]` in `~/.jcode/config.toml` (`refresh_ttl_secs`, `max_results`, `max_files`; zero or absent values fall back to built-in defaults), documented in README and shown in the `/config` settings summary.

## Edge cases covered by tests

- Section-header up-skip wraps correctly for non-power-of-2 list lengths (regression: `wrapping_sub(1) % len` computed `usize::MAX % len`).
- The first-use building-index hint row is rejected by the accept guard and filtered from Tab-completion targets.
- Out-of-range popover selection clamps, all-header lists terminate, empty lists are safe.
- Config: absent section, partial section, zero values, TOML round-trip, and cap enforcement all fall back to documented defaults.
- Index behavior: binary-file filtering, gitignore loading/matching, empty-dir queries, lazy ignored-dir scan.

## Outside-workspace files (`@~` and `@/`)

Queries starting with `~` or `/` bypass the workspace index and read the parent directory live: `@~/...` expands `$HOME` and `@/...` is absolute. Entries skip symlinks, skipped dirs, and dotfiles unless the partial starts with a dot. Chips keep the literal `~/...` display string and expand the tilde at prompt-build time, so an outside-workspace file attaches exactly like a workspace file. FS queries never touch the frecency store, so browsing never pollutes ranking history. Tab-completion is guarded against sentinel-only popovers (no real candidate rows), which previously divided by zero in the cycle math.

## Tradeoffs / assumptions

- frecency state persists per-project on disk (frequency + last-access time), mirroring the approach of editor file pickers; it does not reuse session history because ranking must survive across sessions.
- The CharBag pre-filter is an allocation-light first pass; the regex pass only runs on surviving candidates to bound per-keystroke cost.
- Suggestion reads are memoized per frame and the config accessor is cached, so the per-keystroke path does not touch disk config or re-rank on every render.

## Validation

- `jcode-tui` file_mention/input_ui/suggestions battery: 111 passed at first review; now 164 (89 file_mention + 39 input_ui + 36 suggestions), all green in full-suite order.
- Full `jcode-tui --lib` suite compared against the clean base: failure sets byte-identical (23 pre-existing environment-dependent failures, zero regressions from this PR).
- Root-crate `commands_tests` compile fix included (master commit ff329733f added a `usage` field to `ModelRoute` without updating 5 test initializers, breaking `cargo test -p jcode`; also available standalone on `fix/modelroute-usage-test-init` if preferred as a separate PR).
- `provider_init_tests` env isolation so provider tests are deterministic across developer machines.
- Workspace `cargo check --all-targets` clean; clippy clean for all files touched by this PR; runtime smoke test on a standalone socket passes.

### CODE REVIEW

**Comment by alecuba16** [comment]
Addressed all 5 Greptile findings in follow-up commits (83c4ac7e8..108b13b58):

1. **UTF-8 panic in `common_prefix`** — now operates on char vectors and builds the prefix from chars. Regression test covers éx/éy, café/x.rs, and divergence inside a multi-byte char.
2. **@mention stays active after a space** — `extract_at_query` now requires the @token to extend to the end of the input, so typing `@main then continue` no longer leaves a stale mention intercepting Enter. New tests pin the boundary cases.
3. **Undo loses the file chip** — Ctrl+Z snapshots (`InputUndoEntry`) now carry `file_chips`, so undoing a backspace restores the attachment instead of silently dropping it at send time. Test added.
4. **Full read before size check** — chips now stat the file first and use a bounded read (`read_to_string_bounded`); oversized files are never fully loaded. Tests cover oversize truncation and the read cap.
5. **Frecency leaks across projects** — the log is now keyed by the working dir (`state_dir/file_frecency/<sha256(cwd)>/file_frecency.jsonl`) and `check_refresh` swaps stores when the cwd changes. Isolation test added.

**Comment by alecuba16** [comment]
Fixed the follow-up P1 (`Corrupt Files Lose Context`) in 8d5a0ee33.

`read_to_string_bounded` now distinguishes the two cases explicitly:
- **Byte limit slicing a multi-byte char at the cut** (file itself is valid): keep the truncated prefix.
- **Invalid UTF-8 anywhere before the limit** (genuine corruption): return `InvalidData`, which flows into the existing `[read failed: …]` marker in the prompt.

Tests added: mid-file invalid byte rejected at the reader level, sliced-char-at-limit prefix kept, and end-to-end `build_prompt_with_files` asserting the failure marker appears and no partial content is embedded.

**Comment by alecuba16** [comment]
Addressed the review finding in `filesystem_candidates` (`7cd6c699a`):

- **Ordering defect**: the candidate cap was applied during readdir, so a large directory kept an arbitrary readdir-order subset instead of the lexically first entries. Now all matching names are collected, sorted, then truncated. Regression test `filesystem_candidates_truncate_keeps_lexically_first_entries` reproduces the reported scenario (reverse-inserted entries, cap 3) and asserts `alpha/beta/xray`.
- **Scan cost**: the prefix and dotfile filters now run on the entry name before `file_type()`, so a non-matching query never stats every entry. The symlink/SKIP_DIRS/directory checks still run on the surviving entries only, preserving the original semantics exactly (SKIP_DIRS filtering stays directory-only).
- `test_remote_fallback_provider_suggestions_normalize_bare_openai_openrouter_routes` fails identically on the clean base (machine-config-dependent), passes in full-suite order; not from this PR.

91/91 file_mention tests green; full-suite failure set byte-identical to the pre-PR baseline.

File: crates/jcode-tui/src/tui/app/file_mention.rs [hunk 1]
```diff
@@ -0,0 +1,2971 @@
...
+            };
+            matches.push(FileMatch {
+                score: 0.0,
+                path: display.into(),
+                is_directory,
+                is_recent: false,
+                is_likely_binary,
+            });
+        }
+        matches.sort_by(|a, b| a.path.cmp(&b.path));
```
[the remark this answers is not in the document]
Comment (alecuba16) [reply crates/jcode-tui/src/tui/app/file_mention.rs]: Fixed in `7cd6c699a`: eligible entries are collected first, sorted, then truncated after the sort (`names.sort(); names.truncate(limits.max_results)`). The readdir-order `matches.len() >= limits.max_results` break is gone, so capped results are deterministic and lexically first. Covered by `filesystem_candidates_truncate_keeps_lexically_first_entries` (reverse-inserted entries, cap 3, asserts `alpha`/`beta`/`xray`).

File: crates/jcode-tui/src/tui/app/file_mention.rs [hunk 2]
```diff
@@ -0,0 +1,2971 @@
...
+                continue;
+            }
+            if !partial.is_empty() && !name.starts_with(partial) {
+                continue;
+            }
+            if name.starts_with('.') && !partial.starts_with('.') {
+                continue;
+            }
+            if matches.len() >= limits.max_results {
+                break;
```
[the remark this answers is not in the document]
Comment (alecuba16) [reply crates/jcode-tui/src/tui/app/file_mention.rs]: Fixed in `7cd6c699a`: the prefix and dotfile filters now run on the entry name before `file_type()`, so a non-matching query never stats the whole directory (symlink/SKIP_DIRS checks run only on survivors, preserving original directory-only semantics). Capped results are also deterministic now: collect, sort, then truncate. Two regression tests: `filesystem_candidates_truncate_keeps_lexically_first_entries` and `filesystem_candidates_nonmatching_prefix_returns_empty` (2000-entry dir). 91/91 file_mention tests green.

File: crates/jcode-tui/src/tui/app/state_ui_input_helpers.rs [hunk 3]
```diff
@@ -1589,10 +1746,80 @@ impl App {
...
+            && common.starts_with(&query)
+            && let Some((new_input, new_cursor)) =
+                crate::tui::ui::input_ui::tab_complete(&self.input, &common)
+        {
+            self.remember_input_undo_state();
+            self.input = new_input;
+            self.cursor_pos = new_cursor;
+            self.tab_completion_state = None;
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (alecuba16) [reply crates/jcode-tui/src/tui/app/state_ui_input_helpers.rs:1805]: Fixed on `5baf112f9`: `common_prefix` now counts chars and collects the prefix from `first_chars[..end]`, so multi-byte UTF-8 sequences are never sliced mid-character. Regression tests cover `["éx", "éy"] → "é"` and `["café/x.rs", "café/y.rs"] → "café/"`.

File: crates/jcode-tui/src/tui/ui_input.rs [hunk 4]
```diff
@@ -55,10 +73,115 @@ fn shell_mode_hint(mode: ComposerMode) -> Option<&'static str> {
...
+        return Some((pos, String::new()));
+    }
+
+    let query_len = after_at.chars().take_while(|c| !c.is_whitespace()).count();
+    if query_len == 0 {
+        return None;
+    }
+
+    let query: String = after_at.chars().take(query_len).collect();
+    Some((pos, query))
```
[the remark this answers is not in the document]
Comment (alecuba16) [reply crates/jcode-tui/src/tui/ui_input.rs]: Fixed on `5baf112f9`: `extract_at_query` now rejects the token if any whitespace follows the `@`, so mention mode only stays active while the token extends to the end of the input. Typing past a mention submits the composer instead of accepting a stale suggestion.

File: crates/jcode-tui/src/tui/app/file_mention.rs [hunk 5]
```diff
@@ -0,0 +1,2414 @@
...
+                } else {
+                    let line_count = content.lines().count();
+                    let preview: String = content.lines().take(200).collect::<Vec<_>>().join("\n");
+                    format!(
+                        "{}\n\n[... file too large: {} lines, {} bytes, showing first 200 lines]",
+                        preview,
+                        line_count,
+                        content.len(),
+                    )
+                };
```
[the remark this answers is not in the document]
Comment (alecuba16) [reply crates/jcode-tui/src/tui/app/file_mention.rs]: Addressed on `5baf112f9`: files over 100 KB are skipped via a metadata precheck before any read, each read is bounded by `take(MAX_FILE_SIZE)`, and the cumulative 500 KB budget emits explicit `[skipped: context budget exhausted]` markers. Reads stay synchronous by design but are capped at 100 KB per file.

File: crates/jcode-tui/src/tui/app/input.rs [hunk 6]
```diff
@@ -2774,15 +2774,95 @@ pub(super) fn handle_enter(app: &mut App) -> bool {
...
+                let drain_start =
+                    file_chip_backspace_start(&app.input, app.cursor_pos, &app.file_chips)
+                        .unwrap_or(prev);
                 app.remember_input_undo_state();
-                app.input.drain(prev..app.cursor_pos);
-                app.cursor_pos = prev;
+                app.input.drain(drain_start..app.cursor_pos);
+                app.cursor_pos = drain_start;
                 app.reset_tab_completion();
                 app.sync_model_picker_preview_from_input();
```
[the remark this answers is not in the document]
Comment (alecuba16) [reply crates/jcode-tui/src/tui/app/input.rs:2922]: Fixed on `5baf112f9`: `InputUndoEntry` now snapshots `file_chips` alongside text and cursor, and undo restores them together, so a restored mention keeps its attachment.

File: crates/jcode-tui/src/tui/app/file_mention.rs [hunk 7]
```diff
@@ -0,0 +1,2414 @@
...
+            None
+        } else {
+            Some(crate::storage::durable_state_dir().join("file_frecency.jsonl"))
+        };
+        let mut frecency = Self {
+            data: HashMap::new(),
+            file,
+            cwd: cwd.to_path_buf(),
+        };
+        frecency.load();
```
[the remark this answers is not in the document]
Comment (alecuba16) [reply crates/jcode-tui/src/tui/app/file_mention.rs:268]: Fixed on `5baf112f9`: frecency state is keyed per working directory, so relative-path boosts no longer leak across projects. Covered by `frecency_state_file_is_keyed_by_cwd`.

File: crates/jcode-tui/src/tui/app/file_mention.rs [hunk 8]
```diff
@@ -0,0 +1,2561 @@
...
+    match String::from_utf8(buf) {
+        Ok(s) => Ok(s),
+        Err(e) => {
+            let valid_up_to = e.utf8_error().valid_up_to();
+            if valid_up_to > 0 {
+                // Invalid bytes only affect the tail (e.g. the cut landed
+                // mid-character at the limit): keep the valid prefix.
+                let bytes = e.into_bytes();
+                let s = String::from_utf8_lossy(&bytes[..valid_up_to]).into_owned();
+                Ok(s)
```
[the remark this answers is not in the document]
Comment (alecuba16) [reply crates/jcode-tui/src/tui/app/file_mention.rs]: Fixed on `5baf112f9`: `read_to_string_bounded` only accepts a lossy prefix when the invalid byte sits exactly at the 100 KB read boundary (truncated multi-byte char); interior corruption returns `InvalidData` and surfaces the existing read-failure marker instead of silently dropping the remainder.

File: crates/jcode-tui/src/tui/app/input.rs [hunk 9]
```diff
@@ -3803,6 +3885,22 @@ impl App {
...
+            // Resolve relative paths against workspace root so file reads work
+            // regardless of the process current directory.
+            let cwd: PathBuf = self
+                .session
+                .working_dir
+                .as_deref()
+                .map(PathBuf::from)
+                .unwrap_or_else(|| PathBuf::from("."));
+            input =
+                crate::tui::app::file_mention::build_prompt_with_files(&input, &file_chips, &cwd);
```
[the remark this answers is not in the document]
Comment (alecuba16) [reply crates/jcode-tui/src/tui/app/input.rs]: Fixed on `5baf112f9`: chip expansion now happens in `take_prepared_input`, so all prepare paths (including `queue_message_for_reconnect`) expand and consume the chips, and `PreparedInput` carries them for restore. Regression test `queue_for_reconnect_expands_and_consumes_file_chips` covers the reconnect queue specifically.

**Comment by alecuba16** [comment]
Closing per the triage guidance on #570 — this needs a maintainer decision before landing. The branch (`select_files`, head `5baf112f9`) stays available as reference material for whichever direction is chosen. Happy to reopen if the direction lands on this approach.
